### PR TITLE
Speedups2

### PR DESCRIPTION
**Pull request by proteneer**
This PR:

1. Disables drop connect
2. Restores the 0.5 prefactor to the exp loss function
3. Speeds-up predictions with a queue
4. Adds a test for gdb10 (though only for <23 atoms)
5. Uses RandomSplit as opposed to RandomGroupSplit (just like in the paper)

### CODE REVIEW

File: deepchem/data/datasets.py [hunk 1]
```diff
@@ -737,7 +736,11 @@ def iterbatches(self,
                   batch_size=None,
                   epoch=0,
                   deterministic=False,
-                  pad_batches=False):
+                  pad_batches=False,
+                  dataset_size=None,
+                  load_ys=True,
```
Comment (timaro) [inline deepchem/data/datasets.py:741]: add the arg meanings to the docstring?

File: deepchem/data/datasets.py [hunk 2]
```diff
@@ -857,36 +872,26 @@ def iterate(dataset, batch_size):
...
-        #   np.random.shuffle(w)
-        #   np.random.shuffle(ids)
-        #   tot_permute_time += time.time() - p_time
-          # sample_perm = np.random.permutation(n_shard_samples)
-        # else:
-          # sample_perm = np.arange(n_shard_samples)
+
+        if not deterministic:
+          pp = np.random.permutation(len(X))
+          X = X[pp]
```
Comment (timaro) [inline deepchem/data/datasets.py:880]: Are X and y ever used?

File: deepchem/models/tensorgraph/layers.py [hunk 3]
```diff
@@ -2766,6 +2766,23 @@ def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
     return out_tensor
 
 
+class RootMeanError(Layer):
```
Comment (timaro) [inline deepchem/models/tensorgraph/layers.py:2772]: rather than have this and RootWeightedMeanError (below), why not make a single RootMeanError with optional weights?

File: deepchem/models/tensorgraph/models/symmetry_function_regression.py [hunk 4]
```diff
@@ -94,6 +92,8 @@ def default_generator(self,
                         epochs=1,
                         predict=False,
                         pad_batches=True):
+
```
Comment (timaro) [inline deepchem/models/tensorgraph/models/symmetry_function_regression.py:95]: whitespace-only changes?

File: deepchem/models/tensorgraph/models/symmetry_function_regression.py [hunk 5]
```diff
@@ -364,13 +364,13 @@ def false_fn():
...
-    self.weights = Weights(shape=(None, self.n_tasks))
+    # self.weights = Weights(shape=(None, self.n_tasks))
 
-    loss = RootWeightedMeanError(in_layers=[all_cost, self.weights])
+    loss = RootMeanError(in_layers=[all_cost])
 
     if self.shift_exp:
       print("WARNING: Using shifted exponential loss")
-      loss = ShiftedExponential(1.0, in_layers=loss)
+      loss = ShiftedExponential(0.5, in_layers=loss)
```
Comment (timaro) [inline deepchem/models/tensorgraph/models/symmetry_function_regression.py]: magic numbers in code are best put in constants or configs.

File: deepchem/models/tensorgraph/models/symmetry_function_regression.py [hunk 6]
```diff
@@ -402,29 +402,29 @@ def featurize(self, dataset, feat_dir):
     # (ytz): Featurization uses significantly more memory than training
     # so the batch_size should be smaller
     batch_size = self.batch_size
-    shard_size = batch_size * 3
+    shard_size = batch_size * 512
```
Comment (timaro) [inline deepchem/models/tensorgraph/models/symmetry_function_regression.py:434]: cf. previous comment

File: deepchem/models/tensorgraph/models/symmetry_function_regression.py [hunk 7]
```diff
@@ -520,26 +491,36 @@ def feed_queue():
       pool.close()
       t1.join()
 
-    return dc.data.DiskDataset.create_dataset(
+    res = dc.data.DiskDataset.create_dataset(
         shard_generator=shard_generator(self),
         data_dir=feat_dir,
         X_is_sparse=True)
 
+    # print(res.get_shape(), dataset.get_shape())
```
Comment (timaro) [inline deepchem/models/tensorgraph/models/symmetry_function_regression.py:528]: just delete the commented lines and get rid of res?

File: deepchem/models/tensorgraph/models/symmetry_function_regression.py [hunk 8]
```diff
@@ -520,26 +491,36 @@ def feed_queue():
...
 
   def default_generator(self,
                         dataset,
                         epochs=1,
                         predict=False,
                         deterministic=True,
-                        pad_batches=True):
+                        pad_batches=False):
+
+    assert pad_batches == False
```
Comment (timaro) [inline deepchem/models/tensorgraph/models/symmetry_function_regression.py:540]: lol. remove pad_batches as an option?

File: deepchem/models/tensorgraph/models/symmetry_function_regression.py [hunk 9]
```diff
@@ -520,26 +491,36 @@ def feed_queue():
...
 
     if predict and not deterministic:
       raise Exception(
             "Called predict with non-deterministic generator, terrible idea.")
 
+    # print("DEFAULT_GENERATOR BATCH_SIZE", self.batch_size)
+
     batch_size = self.batch_size
 
+    # way faster on un-featurized data
```
Comment (timaro) [inline deepchem/models/tensorgraph/models/symmetry_function_regression.py:550]: vs what?

File: deepchem/models/tensorgraph/models/symmetry_function_regression.py [hunk 10]
```diff
@@ -549,17 +530,42 @@ def default_generator(self,
       print("Requires manual featurization")
       pass
 
+
+    # print("??")
```
Comment (timaro) [inline deepchem/models/tensorgraph/models/symmetry_function_regression.py:563]: these lines would seem to be worth deleting.

File: deepchem/models/tensorgraph/models/symmetry_function_regression.py [hunk 11]
```diff
@@ -549,17 +530,42 @@ def default_generator(self,
...
+      if predict:
+        generator = dataset.iterbatches(
+          deterministic=deterministic,
+          batch_size=batch_size,
+          dataset_size=dataset_size,
+          load_ys=False,
+          load_ws=False,
+          load_ids=False)
+      else:
+        generator = dataset.iterbatches(
```
Comment (timaro) [inline deepchem/models/tensorgraph/models/symmetry_function_regression.py:577]: unnecessary duplication. why not:

generator = dataset.iterbatches(...
                                                      load_ys=!predict,
                                                      ...)

?

File: deepchem/models/tensorgraph/models/symmetry_function_regression.py [hunk 12]
```diff
@@ -549,17 +530,42 @@ def default_generator(self,
...
+        # print(YIELDING)
 
+        # print("X_b shape", X_b.shape)
         feed_dict = {}
         if y_b is not None and not predict:
           for index, label in enumerate(self.labels_fd):
             feed_dict[label] = y_b[:, index:index + 1]
         if w_b is not None and not predict:
-          feed_dict[self.weights] = w_b
+          feed_dict[self.weights] = np.ones((self.batch_size, 1))
```
Comment (timaro) [inline deepchem/models/tensorgraph/models/symmetry_function_regression.py:595]: so, if w_b is set and we're not predicting, the intent is to ignore w_b? and if w_b *is* none, or we are predicting...?

File: deepchem/models/tensorgraph/symmetry_functions.py [hunk 13]
```diff
@@ -401,12 +401,15 @@ def __init__(self,
                init='ani_norm',
                # activation='relu',
                activation='gaussian',
+               drop_connect=False,
```
Comment (timaro) [inline deepchem/models/tensorgraph/symmetry_functions.py:404]: this appears to be unused now? remove?

File: deepchem/models/tensorgraph/symmetry_functions.py [hunk 14]
```diff
@@ -432,9 +435,13 @@ def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
     # The second implements drop connect, whereby the weights are dropped
     # out randomly. ytz thinks that clipping then dropping is the correct
     # order of operations. ytz may also be wrong.
-    self.W = tf.clip_by_norm(self.W, 1.0, axes=1)
-    drop_p = 0.5
-    self.W = tf.nn.dropout(self.W, keep_prob=drop_p) * drop_p
+    self.W = tf.clip_by_norm(self.W, 3.0, axes=1)
```
Comment (timaro) [inline deepchem/models/tensorgraph/symmetry_functions.py]: why 3.0 instead of 1.0? magic number.

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 15]
```diff
@@ -295,11 +286,23 @@ def fit_generator(self,
         # print("tmp loss:", fetched_values[1])
         n_averaged_batches += 1
         self.global_step += 1
+
+        if (batch_idx % (max_batches//25)) == 0 or batch_idx == max_batches - 1:
+          # print(avg_loss)
+          tmp_loss = float(avg_loss) / n_averaged_batches
+          avg_speed = ((self.global_step - start_step) * self.batch_size /
+                       (time.time() - start_time)) * 60
+          t_time = time.time() - start_time
```
Comment (timaro) [inline deepchem/models/tensorgraph/tensor_graph.py:363]: redundant time.time() - start_time calls w/previous line means that time delta will differ (probably not important here), but also extra code. move this line up, then use t_time in avg_speed calc

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 16]
```diff
@@ -316,6 +319,7 @@ def fit_generator(self,
                                                             avg_loss))
         saver.save(self.session, self.save_file, global_step=self.global_step)
         time2 = time.time()
+
```
Comment (timaro) [inline deepchem/models/tensorgraph/tensor_graph.py:390]: do you not have your editor configured to strip trailing spaces?

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 17]
```diff
@@ -370,11 +374,16 @@ def default_generator(self,
           feed_dict[initial_state] = zero_state
         yield feed_dict
 
-  def predict_on_generator(self, generator, transformers=[], outputs=None):
+  def predict_on_generator(
+    self,
+    pred_generator,
+    transformers=[],
+    outputs=None,
```
Comment (timaro) [inline deepchem/models/tensorgraph/tensor_graph.py:449]: document these params?

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 18]
```diff
@@ -388,38 +397,72 @@ def predict_on_generator(self, generator, transformers=[], outputs=None):
...
+      for iiii in range(max_batches):
+        feed_results = self.session.run(
+          outputs,
+          feed_dict={self._training_placeholder: 0.0})
+        results.append(feed_results[0])
+
+      final_results = np.concatenate(results)
+      return final_results
+
+      # final_results = []
```
Comment (timaro) [inline deepchem/models/tensorgraph/tensor_graph.py]: oy...it would really be better to just make a new method with the new behavior and move this old code to a new method name with a docstring explaining its use and existence, if the intent is to keep it around for some reason.

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 19]
```diff
@@ -867,6 +914,34 @@ def load_from_dir(model_dir):
   def __del__(self):
     pass
 
+def _enqueue_batch_pred(tg, generator, graph, sess):
+  """
+  Function to load data into
+  Parameters
+  ----------
+  tg
```
Comment (timaro) [inline deepchem/models/tensorgraph/tensor_graph.py:1053]: what are the expected types and the parameter meanings?

File: deepchem/models/tensorgraph/tensor_graph.py [hunk 20]
```diff
@@ -886,10 +961,15 @@ def _enqueue_batch(tg, generator, graph, sess, n_enqueued, final_sample):
...
     for feed_dict in generator:
       enq = {}
+
+      # print("SETTING", tg._training_placeholder, "TPH 1.0")
       enq[tg._training_placeholder] = 1.0
-      for layer in tg.features + tg.labels + tg.task_weights:
+      for layer in tg.features + tg.labels:
         obj = feed_dict[layer]
         enq[tg.get_pre_q_input(layer).out_tensor] = obj
+      for layer in tg.task_weights:
```
Comment (timaro) [inline deepchem/models/tensorgraph/tensor_graph.py:1101]: you should add an explanatory comment here.

File: examples/roitberg/roitberg.py [hunk 21]
```diff
@@ -1,9 +1,9 @@
 #!/usr/bin/env python
-# import cProfile
-# pr = cProfile.Profile()
-# pr.disable()
-# import pstats
-# from io import StringIO
+import cProfile
```
Comment (timaro) [inline examples/roitberg/roitberg.py]: cProfile slows things down. probably not a good idea to check this in.

File: examples/roitberg/roitberg.py [hunk 22]
```diff
@@ -110,7 +113,14 @@ def find_training_data(base_dir, max_gdb_level):
 
   return files
 
-def load_roitberg_ANI(args, mode="atomization"):
+def find_gdb10_test_data(base_dir):
+  base_dir = os.path.abspath(os.path.expanduser(base_dir))
+  file = os.path.join(base_dir, "ani1_gdb10_ts.h5")
```
Comment (timaro) [inline examples/roitberg/roitberg.py]: file = path(base_dir, "ani1_gdb10_ts.h5")

(assuming you've merged in my previous changes)

File: examples/roitberg/roitberg.py [hunk 23]
```diff
@@ -133,12 +143,13 @@ def load_roitberg_ANI(args, mode="atomization"):
...
 
-  hdf5files = find_training_data(args.training_data_dir, args.gdb_level)
+  # hdf5files = find_training_data(args.training_data_dir, args.gdb_level)
+  # hdf5files.append(find_gdb10_test_data(args.training_data_dir))
   groups = []
 
   def shard_generator():
 
-    shard_size = 4096 * args.batch_size
+    shard_size = 4096 * batch_size
```
Comment (timaro) [inline examples/roitberg/roitberg.py]: magic number should be a constant

File: deepchem/data/datasets.py (hunk 2 above)
Comment (proteneer) [reply deepchem/data/datasets.py:880]: Yes they're used in the while loop below

File: deepchem/data/datasets.py (hunk 1 above)
Comment (proteneer) [reply deepchem/data/datasets.py:741]: I'm not calling this function anymore

File: deepchem/models/tensorgraph/models/symmetry_function_regression.py (hunk 8 above)
Comment (proteneer) [reply deepchem/models/tensorgraph/models/symmetry_function_regression.py:540]: I will take care of this in the merge upstream

File: deepchem/models/tensorgraph/models/symmetry_function_regression.py (hunk 7 above)
Comment (proteneer) [reply deepchem/models/tensorgraph/models/symmetry_function_regression.py:528]: Comment cleanup will come after the big merge
